core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -3035,6 +3001,86 @@ private void loadConfig() throws IOException {
             // load from disk
             cfg.unmarshal(Jenkins.this);
         }
+
+        try {
+            checkRawBuildsDir(buildsDir);
+            setBuildsAndWorkspacesDir();
+        } catch (InvalidBuildsDir invalidBuildsDir) {
+            throw new IOException(invalidBuildsDir);
+        }
+
+    }
+
+    private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
+        boolean mustSave = false;
+        String newBuildsDir = SystemProperties.getString(BUILDS_DIR_PROP);
+        if (newBuildsDir != null && !buildsDir.equals(newBuildsDir)) {
+
+            checkRawBuildsDir(newBuildsDir);
+            LOGGER.log(Level.WARNING, "Changing builds directories from {0} to {1}. Beware that no automated data migration will occur.",
+                       new String[]{buildsDir, newBuildsDir});
+            buildsDir = newBuildsDir;
+            mustSave = true;
+        } else if (!isDefaultBuildDir()) {
+            LOGGER.log(Level.INFO, "Using non default builds directories: {0}.", buildsDir);
+        }
+
+        String newWorkspacesDir = SystemProperties.getString(WORKSPACES_DIR_PROP);
+        if (newWorkspacesDir != null && !workspaceDir.equals(newWorkspacesDir)) {
+            LOGGER.log(Level.WARNING, "Changing workspaces directories from {0} to {1}. Beware that no automated data migration will occur.",
+                       new String[]{workspaceDir, newWorkspacesDir});
+            workspaceDir = newWorkspacesDir;
+            mustSave = true;
+        } else if (!isDefaultWorkspaceDir()) {
+            LOGGER.log(Level.INFO, "Using non default workspaces directories: {0}.", workspaceDir);
+        }
+
+        if (mustSave) {
+            save();
+        }
+    }
+
+    /**
+     * @param value
+     */
+    @VisibleForTesting
+    /*private*/ static void checkRawBuildsDir(String value) throws InvalidBuildsDir {
+
+        // do essentially what expandVariablesForDirectory does, without an Item
+        String replacedValue = expandVariablesForDirectory(value,
+                                                           "doCheckRawBuildsDir-Marker:foo",
+                                                           Jenkins.getInstance().getRootDir().getPath() + "/jobs/doCheckRawBuildsDir-Marker$foo");
+
+        File replacedFile = new File(replacedValue);
+        if (!replacedFile.isAbsolute()) {
+            throw new InvalidBuildsDir(value + " does not resolve to an absolute path");
+        }
+
+        if (!replacedValue.contains("doCheckRawBuildsDir-Marker")) {
+            throw new InvalidBuildsDir(value + " does not contain ${ITEM_FULL_NAME} or ${ITEM_ROOTDIR}, cannot distinguish between projects");
+        }
+
+        if (replacedValue.contains("doCheckRawBuildsDir-Marker:foo")) {
+            // make sure platform can handle colon
+            try {
+                File tmp = File.createTempFile("Jenkins-doCheckRawBuildsDir", "foo:bar");
```

Comment:
this would result in an illegal filename on windows as `:` is a reserved character - how are the tests passing??

Reply:
@jtnord no idea. Can check, but this test is a cut-n-paste from before, and looking around here, it does not seem it was in a test using `assume`. Checking on CI side.

Reply:
I found it really hard to even grok what this whole method was trying to do. 
if it is cut and paste then _shrug_

Reply:
We're trying to figure out whether there would be legal file names for a given item name (without having an item). This particular part implements the `ITEM_FULLNAME` vs. `ITEM_FULL_NAME` compatibility check.

Reply:
So, didn't dig into _why_ it works, but this test, from `JenkinsDescriptorTest` **is** currently successful in CI on master: https://ci.jenkins.io/job/Core/job/jenkins/job/master/867/testReport/jenkins.model/JenkinsDescriptorTest/Windows___Windows_Publishing___testBuildDirValidation/

![image](https://user-images.githubusercontent.com/223853/38996218-0ca4b09e-43eb-11e8-8bd3-16a64cf9d723.png)

Reply:
@batmat There is no test for `ITEM_FULLNAME` due to the platform dependency, so this doesn't even happen during the test. This used to be basic form validation, so I didn't bother.

What's the problem here anyway? If the user wants `ITEM_FULLNAME`, that's only OK if the file system can handle colons. That's all this check does. Otherwise the error message will tell them to use `ITEM_FULL_NAME`, which escapes colons.